File: Python/Problem-005.py
#
# Solution to Problem 5: Smallest Multiple
# by Bill Dybas
#

def is_divisible(n):

    # Divisible by 12 means also divisible by 3 and 4
    # Divisible by 14 means also divisible by 2 and 7
    # Divisible by 15 means also divisible by 3 and 5
    # Divisible by 16 means also divisible by 2, 4, and 8
    # Divisible by 18 means also divisible by 2, 3, 6, and 9
    # Divisible by 20 means also divisible by 2, 4, 5, and 10
    if not n % 11 == 0 or not n % 12 == 0 or not n % 13 == 0 \
        or not n % 14 == 0 or not n % 15 == 0 or not n % 16 == 0 \
        or not n % 17 == 0 or not n % 18 == 0 or not n % 19 == 0 \
        or not n % 20 == 0:
        return False
    else:
        return True

def calculate():
    result = 1
    while not is_divisible(result):
        result += 1

    return result

if __name__ == '__main__':
    # Brute Force Method by checking divisibility
    print(calculate())

    # Could also note that:
    # Result = LCM({2..20})
    # Result = 2 * 3 * 2 * 5 * 7 * 3 * 11 * 13 * 2 * 17 * 19

File: JavaScript/Problem-005.js
//
// Solution to Problem 5: Smallest Multiple
// by Bill Dybas
//

// See: https://en.wikipedia.org/wiki/Greatest_common_divisor#Using_Euclid.27s_algorithm
const gcd = function(a, b) {
  return !b ? a : gcd(b, a % b);
}

// See: https://en.wikipedia.org/wiki/Least_common_multiple#Reduction_by_the_greatest_common_divisor
const lcm = function(a, b) {
  return Math.abs(a) / gcd(a, b) * Math.abs(b);
}

const answer =
  Array(20)
    .fill() // Fill an Array w/ 0's
    .map((x, i) => i + 1) // Make the index + 1 the value at that index
    .reduce((previous, current) => lcm(previous, current)); // Find the LCM of all the numbers

console.log(answer);
